lexicon/providers/gratisdns.py

```diff
@@ -0,0 +1,278 @@
+
+    # List all records. Return an empty list if no records found
+    # type, name and content are used to filter records.
+    # If possible filter during the query, otherwise filter after response is received.
+    # pylint: disable=too-many-locals
+    def _list_records(self, rtype=None, name=None, content=None):
+        query_params = {'action': 'dns_primary_changeDNSsetup',
+                        'user_domain': self.domain_id}
+        response = self._get(query_params=query_params)
+        html = BeautifulSoup(response.content, "html.parser")
+        # List of records
+        all_records = []
+        # DNS records are grouped by type
+        dns_record_types = html.find_all("div", 'dns-records')
+        for dns_record in dns_record_types:
+            # Find the title of this group, aka record type
+            record_type = dns_record.find("h2").contents[0].strip()
+            # The entries themselves follow in the table below
+            # Each row is a single entry
+            entry_table_rows = dns_record.find('tbody').find_all('tr')
+            for row in entry_table_rows:
+                cols = row.find_all('td')
+                # Format is; name, content, ttl, EDIT/DELETE button
+                record_name = cols[0].contents[0]
+                record_content = cols[1].contents[0]
+                record_ttl = cols[2].contents[0]
+                # We have to pull the ID from the button link
+                record_id = None
+                button_link = cols[3].find('a')
+                if button_link:
+                    start = button_link['href'].index('&id=')+4
+                    end = button_link['href'].index('&', start)
+                    record_id = button_link['href'][start:end]
+
+                processed_record = {
+                    'type': record_type,
+                    'name': record_name,
+                    'ttl': record_ttl,
+                    'content': record_content,
+                    'id': record_id,
+                }
+                all_records.append(processed_record)
+
+        records = self._filter_records(
+            records=all_records,
+            rtype=rtype,
+            name=name,
+            content=content
+        )
+        LOGGER.debug('list_records: %s', records)
+        return records
+
+    # Filter a list of records based on criteria
+    # NOTE: Duplicate of lexicon/providers/transip.py _filter_records
+    def _filter_records(self, records, rtype=None, name=None, content=None):
+        _records = []
+        for record in records:
+            if ((not rtype or record['type'] == rtype) and  # pylint: disable=too-many-boolean-expressions
+                    (not name or self._full_name(record['name']) == self._full_name(name)) and
+                    (not content or record['content'] == content)):
+                _records.append(record)
+        return _records
+
+    # pylint: disable=no-self-use
+    def _get_content_entry(self, rtype):
```

Comment:
Not sure exactly what this is doing. What about other record types that lexicon supports (will support in the future)?

Reply:
Are these the only record types that GratisDNS supports?

Reply:
1. SOA entries are supported, but:
     * It is handled completely in a completely different way, compared to the rest of the records.

Support could be added, but it will be a hassle.

2. MX entries are supported, but;
    * The entries are handled differently from A, AAA, CNAME and TXT.

I will consider adding support for it

3. NS entries are supported, but:
    * The default entries cannot be removed. I did not know how to deal with this with regards to the interface.
    * The default entries are ID-less. I did not know how to deal with this with regards to the interface.

Assuming we find a way to deal with these issues, I will consider adding support.

4. SPF entries are not supported.
    * Should I simply create an equivalent TXT record?
    * Should TXT records which contain SPF entries be presented as SPF?

Assuming we find a way to deal with these issues, I will consider adding support.

5. SRV entries are supported, but:
    * The entries are handled differently from A, AAA, CNAME and TXT.

I will consider adding support for it

6. LOC entries are not supported.

---

Several other entries are however supported, assuming they will be added down the line, they are:
* CAA (Certification Authority Authorization)
* TLSA
* DS
* SSHFP (Secure SHell Finger Print)
* AFSDB (Andrew File System Data Base)

---

Should we consider adding tests for the different entries (MX, NS, SPF, SRV, etc)?

Reply:
Ah understood. 
I'd like to see the common record types supported by default, but you are correct, we haven't explicitly tested the providers to ensure they handle the required record types.

Here's the list of records I'd like to have supported in every provider:
https://github.com/AnalogJ/lexicon/blob/master/lexicon/parser.py#L19

`'A', 'AAAA', 'CNAME', 'MX', 'NS', 'SOA', 'TXT', 'SRV', 'LOC'`

I'd be willing to merge this provider as is, but would you be willing to work on adding support for the remaining record types in a separate PR?

Reply:
Yes, I can work on adding the remaining record types, assuming we figure out how to handle the issues I presented above. Specifically what should be done with the default NX entries, SPF entries and the missing LOC entry support.

Reply:
@AnalogJ can you throw your thoughts about the above whenever it is convinient for you?